### PR TITLE
Wayland: Multilanguage keybindings support

### PR DESCRIPTION
**Pull request by dydyamotya**
I'm from Russia, so I needed keybindings on two key layouts. Cause I'm trying to use Wayland, I've written code for Wayland. Look, pls, if it satisfy your requirements. I think these should work for X11 also. If so, I can do this work.

### CODE REVIEW

File: libqtile/backend/wayland/core.py [hunk 1]
```diff
@@ -119,6 +119,7 @@ def __init__(self):
         self.keyboards: List[keyboard.Keyboard] = []
         self.grabbed_keys: List[Tuple[int, int]] = []
         self.grabbed_buttons: List[Tuple[int, int]] = []
+        self.multilang_mapping = MultilanguageMapping()
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:122]: Please could you rename this `self._multilang_mapping` as it's only used by this class.

File: libqtile/backend/wayland/core.py [hunk 2]
```diff
@@ -864,3 +869,22 @@ def cmd_change_vt(self, vt: int) -> bool:
...
+                keysyms = tuple(keymap.key_get_syms_by_level(keycode, layout, 0)[0] for layout in range(layouts))
+                for keysym in keysyms:
+                    self.mapping[keysym] = keysyms
+
+    def get_keysyms(self, key: Union[config.Key, config.KeyChord]):
+        keysym = xkb.keysym_from_name(key.key, case_insensitive=True)
+        try:
+            return self.mapping[keysym]
+        except KeyError:
+            return (keysym, )
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:890]:
```suggestion
self.mapping.get(keysym, (keysym, ))
```

File: libqtile/backend/wayland/core.py [hunk 3]
```diff
@@ -864,3 +869,22 @@ def cmd_change_vt(self, vt: int) -> bool:
         if not success:
             logger.warning(f"Could not change VT to: {vt}")
         return success
+
+class MultilanguageMapping:
+    def __init__(self):
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:874]: Please could you move this class to `keyboard.py`

File: libqtile/backend/wayland/core.py [hunk 4]
```diff
@@ -748,17 +749,21 @@ def get_screen_info(self) -> List[Tuple[int, int, int, int]]:
...
+        #keysym = xkb.keysym_from_name(key.key, case_insensitive=True)
+        keysyms = self.multilang_mapping.get_keysyms(key)
         mask_key = wlrq.translate_masks(key.modifiers)
-        self.grabbed_keys.append((keysym, mask_key))
-        return keysym, mask_key
+        for keysym in keysyms:
+            self.grabbed_keys.append((keysym, mask_key))
+        return keysyms, mask_key
 
     def ungrab_key(self, key: Union[config.Key, config.KeyChord]) -> Tuple[int, int]:
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:759]: Please update the return type annotation here to reflect the multiple keysyms that could be returned. Same above with `grab_key`

**Review by m-col — CHANGES_REQUESTED** [review]
Nice, this is looking great so thank you for the contribution! I've added some review comments inline but overall it's looking really good.

**Comment by dydyamotya** [comment]
I've changed everything as you suggested.
Have looked also on X11 grab_key implementation, don't know if I'd like to implement it (If someone needs it on X11). Only if you insist :D
Also it should be noted, that my implementation works if XKBCOMMON_DEFAULT variables defined before qtile launching.

There was a discussion about shortcuts. Is it possible to link it here? Or link this PR to it? Sorry, I'm noob at these things.

**Comment by ohadeytan** [comment]
Any news about this? I'm using a Hebrew layout. Currently I'm using [this](https://github.com/qtile/qtile/discussions/2453#discussioncomment-1555818) as a workaround.

**Review by m-col — COMMENTED** [review]
Hey, sorry for forgetting about this. I've rebased and updated it all for you so you don't have to.

I also added another change to convert the style from a class to a function, because we don't need to keep any state around from this code so we can save some space by only getting the mappings as required (i.e. upon key grab).

Overall this looks good but unfortunately I can't test it as I don't have different keyboard layouts. Is that what is required to test it? @ohadeytan would you be willing to test it out too?

**Comment by ohadeytan** [comment]
Hi @m-col, thanks for taking care of that. I tried that out and it didn't work for me, but I'm using X11, so it shouldn't work, isn't it? If it should, I will try to look into it more, maybe I missed some needed configuration.

**Comment by m-col** [comment]
No I'm afraid this applies to Wayland only, sorry

**Review by m-col — APPROVED** [review]

**Comment by m-col** [comment]
No reply on the related issue https://github.com/qtile/qtile/issues/2631, so happy to merge it as is. I've rebased it again and I'll enable it to auto-merge. Thanks for the submission! (and sorry for taking almost a year to merge it!)

**Comment by Freemason-12** [comment]
Hey, why is this not merged? It seems like there's no issue with this PR.

**Pull request closed without merging**